compiler/z/codegen/BinaryEvaluator.cpp

```diff
@@ -2476,56 +2504,75 @@ genericLongAndAsRotateHelper(TR::Node * node, TR::CodeGenerator * cg)
       TR::Node * skipedConversion = NULL;
       TR::Compilation *comp = cg->comp();
 
-      // Transform land into RISBG as it may be better than discrete pair of ands
+      // Given the right case, transform land into RISBG as it may be better than discrete pair of ands
       if (node->getOpCodeValue() == TR::land && secondChild->getOpCode().isLoadConst())
          {
-         uint64_t value = secondChild->getLongInt();
+         uint64_t longConstValue = secondChild->getLongInt();
 
          if ((firstChild->getOpCodeValue() == TR::i2l || firstChild->getOpCodeValue() == TR::iu2l)
                && firstChild->getReferenceCount()==1 && firstChild->getRegister() == NULL && firstChild->isHighWordZero() )
             {
-            skipedConversion = firstChild;
-            firstChild = firstChild->getFirstChild();
-            value &= 0x00000000FFFFFFFFL;
-            }
+            TR::DebugCounter::incStaticDebugCounter(cg->comp(), TR::DebugCounter::debugCounterName(cg->comp(), "codegen/z/BinaryEvaluator/simplifierShouldHaveFoldedThisTree/%s", cg->comp()->signature()));
```

Comment:
This is not true. The tree should have been folded by the simplifier if and only if `value & 0x00000000FFFFFFFFL == 0`. We should not blindly increment the counter here without such a check.

Reply:
So based on our discussion earlier my understanding was that this type of code should be in the optimizer. The optimizer would/should be able to simplify a tree that looks like this, and not just for the specific case where `value & 0x00000000FFFFFFFFL == 0`. So we would leave a check like this here because it could help identify future opportunities. Is this not true?

Reply:
That is only partially correct. Consider the following tree:

```
land
  i2l (highWordZero)
    ...
  lconst 1
```

This tree would enter the above `if` statement and your debug counter would be incremented. However this tree cannot be folded by the simplifier because `value & 0x00000000FFFFFFFFL != 0`, i.e. you have to carry out the AND operation. So the debug counter spots a false positive.

I'm more inclined to say we just remove this entire `if` all together. I don't particularly see much value in it.